### DISCUSSION TITLE
adding mark_in/mark_out options blocks all traffic between roadwarriors

### DISCUSSION THREAD

**Discussion by xnr13** [strongswan/strongswan#2370] [General]
We have a working setup, but now we need to add broadcast between roadwarriors with the help of the forecast plugin. As soon as i add mark_in=%unique and mark_out=%unique to  swanctl.conf, all traffic **between** roadwarriors ceases. At first i thought it was [related to NAT](https://wiki.strongswan.org/issues/3392), so i replicated this setup to a new minimal-config test machine without NAT and can't figure out what is wrong. I attach all the files usually requested for a discussion.

[charon_debug.log.txt](https://github.com/user-attachments/files/16404709/charon_debug.log.txt)
[forecast.conf.txt](https://github.com/user-attachments/files/16404710/forecast.conf.txt)
[ip address.txt](https://github.com/user-attachments/files/16404711/ip.address.txt)
[ip route show table all.txt](https://github.com/user-attachments/files/16404712/ip.route.show.table.all.txt)
[iptables-save.txt](https://github.com/user-attachments/files/16404713/iptables-save.txt)
[swanctl -l.txt](https://github.com/user-attachments/files/16404714/swanctl.-l.txt)
[swanctl -L.2.txt](https://github.com/user-attachments/files/16404814/swanctl.-L.2.txt)
[swanctl.conf.txt](https://github.com/user-attachments/files/16404715/swanctl.conf.txt)

**Comment by tobiasbrunner** [comment]
The plugin installs rules in the OUTPUT and PREROUTING chains to mark packets to the remote IPs when the gateway sends any or forwards traffic from hosts behind it (see [the docs](https://docs.strongswan.org/docs/5.9/plugins/forecast.html) for details - your call of `iptables-save` was apparently after terminating the SAs, so there were no rules).

The rules in PREROUTING also apply to decrypted packets, i.e. they first get the mark of their own tunnel (applied to the ESP packet before decryption) and then the decrypted packet with a roadwarrior's destination address gets tagged with the target mark (you can see that if you use the TRACE target in iptables).

That mark switch is exactly the reason for the observed behavior as it causes the forward policy of the original tunnel to not match anymore because the packet now bears the mark of the target CHILD_SA. And since there is no policy with the target mark that has matching selectors (in the outbound direction) either, you'll see the packet getting dropped and `XfrmInNoPols` increasing in `/proc/net/xfrm_stat`.

I wrote [an analysis of this issue](https://lists.strongswan.org/pipermail/dev/2019-February/001970.html) a few years ago on the old mailing list, where I also explained why the `policies_fwd_out` option, which causes the installation of forward policies in the outbound direction, won't help. The workaround I came up with was to manually install a forward policy that allows the traffic between roadwarrior addresses. In your case something like this:

    ip xfrm policy add src 192.168.111.0/24 dst 192.168.111.0/24 dir fwd priority 50000 tmpl proto esp mode tunnel

No mark is set and the simple template matches packets from any client SA (no restriction on IPs and reqid).  The priority is arbitrary
but it must be higher (numerically lower) than that of any forward policy that's installed with the CHILD_SAs (their priorities are between 200000 and 68541 in current releases). I hope this helps.

**Reply by xnr13** [reply]
> The plugin installs rules in the OUTPUT and PREROUTING chains to mark packets to the remote IPs when the gateway sends any or forwards traffic from hosts behind it (see [the docs](https://docs.strongswan.org/docs/5.9/plugins/forecast.html) for details - your call of `iptables-save` was apparently after terminating the SAs, so there were no rules).

Actually, i tried to be consistent - making sure tunnels were up at the time. Indeed `iptables-save` doesn't return anything, but `iptables-legacy-save` does (attached [iptables-legacy-save.txt](https://github.com/user-attachments/files/16419705/iptables-legacy-save.txt)). Could this be an issue?

> ```
> ip xfrm policy add src 192.168.111.0/24 dst 192.168.111.0/24 dir fwd priority 50000 tmpl proto esp mode tunnel
> ```

Thanks. Indeed after adding this policy unicast connectivity between roadwarriors even with the plugin activated and mark_in=%unique mark_out=%unique is re-established. However, i still can't get any broadcast/multicast through between the roadwarriors, so i'm back to square one before i tried activate the forecast plugin. My only change so far from the config in the original post was adding `remote_ts=dynamic,224.0.0.0/4,192.168.111.255` to swanctl.conf.

Should i also add some ip xfrm policy for multicast 224.0.0.0/4? But your xfrm policy should cover the broadcast to 192.168.111.255, right?

One more observation, when i restart strongswan with `systemctl restart strongswan` or restart one of the tunnels on Roadwarriors, i can briefly see multicasts getting through. I just hope it's not a Roadwarrior issue, but i am testing with firewall off, split-tunneling off and also made sure VPN interface metrics are numerically as low as possible (=2). Roadwarriors, when connected directly to the same LAN, have no issues with exchanging multicasts/broadcasts.

**Reply by tobiasbrunner** [reply]
> Actually, i tried to be consistent - making sure tunnels were up at the time. Indeed `iptables-save` doesn't return anything, but `iptables-legacy-save` does (attached [iptables-legacy-save.txt](https://github.com/user-attachments/files/16419705/iptables-legacy-save.txt)). Could this be an issue?

I see. Since there are no rules configured via nftables (as indicated by your previous post), there shouldn't be an issue.

> My only change so far from the config in the original post was adding `remote_ts=dynamic,224.0.0.0/4,192.168.111.255` to swanctl.conf.

Depending on the direction of the multicast messages, you have to modify `local_ts` as well. And you might have to change the client configuration too. Then check that the traffic selectors are negotiated correctly.

> Should i also add some ip xfrm policy for multicast 224.0.0.0/4? But your xfrm policy should cover the broadcast to 192.168.111.255, right?

No, that's not necessary. But between the clients, such messages are only forwarded by the plugin if you enable `reinject` (see [the docs](https://docs.strongswan.org/docs/5.9/plugins/forecast.html)).

> One more observation, when i restart strongswan with `systemctl restart strongswan` or restart one of the tunnels on Roadwarriors, i can briefly see multicasts getting through.

Are the roadwarriors directly connected to your VPN gateway? Otherwise, how would packets from them even reach the gateway when the SAs are deleted during the restart?

**Reply by xnr13** [reply]
> > One more observation, when i restart strongswan with `systemctl restart strongswan` or restart one of the tunnels on Roadwarriors, i can briefly see multicasts getting through.
> 
> Are the roadwarriors directly connected to your VPN gateway? Otherwise, how would packets from them even reach the gateway when the SAs are deleted during the restart?

What i though was a temporary multicast connectivity turned out to be something else. But answering your question, i had tried both direct connection and externally via a router with port forwarding - picture was same.

Anyway, it took me a while to clarify with the help of wireshark what was happening. It seems all three traffic classes are routed correctly between the roadwarriors now. The culprits were indeed the clients. The problem why i couldn't see some of the multicast messages was in a Windows Filtering Platform hidden filter that was blocking this kind of traffic.
